File: 은지/LV2/프렌즈 4블록.py
def solution(m, n, board):
    b = []
    for r in range(m):
        sub_b = []
        for c in range(n):
            sub_b.append(board[r][c])
        b.append(sub_b)
    
    dr = [0, 1, 0, 1]
    dc = [0, 0, 1, 1]
    answer = 0
    while True:
        chk = [[0]*n for _ in range(m)]
        cnt = 0
        for r in range(m-1):
            for c in range(n-1):
                friends = b[r][c]
                if friends != "0":                    
                    for i in range(1, 4):
                        if b[r+dr[i]][c+dc[i]] != friends:
                            break
                    else:
                        cnt += 1
                        for j in range(4):
                            chk[r+dr[j]][c+dc[j]] = 1

        if cnt == 0:
            break

        for r in range(m):
            for c in range(n):
                if chk[r][c] == 1:
                    answer += 1
                    for k in range(r-1, -1, -1):
                        b[k+1][c] = b[k][c]
                    b[0][c] = "0"
    return answer

m = 5
n = 6
b = ["AAAAAA", "BBAATB", "BBAATB", "JJJTAA", "JJJTAA"]
print(solution(m, n, b))

# print(solution(2, 4, ["baab", "baab"]))

File: 은지/LV2/프렌즈 4블록.js
function solution(m, n, board) {
  var b = []
  for (let r=0; r<m; r++) {
    var sub_b = []
    for (let c=0; c<n; c++) {
      sub_b.push(board[r][c])
    }
    b.push(sub_b)
  }

  var dr = [0, 1, 0, 1]
  var dc = [0, 0, 1, 1]
  var answer = 0;

  while (true) {
    var chk = new Array(m)
    for (let mm=0; mm<m; mm++) {
      chk[mm] = new Array(n)
      for (let nn=0; nn<n; nn++) {
        chk[mm][nn] = 0
      }
    }
    var cnt = 0
    for (let r=0; r<m-1; r++) {
      for (let c=0; c<n-1; c++) {
        var friends = b[r][c]
        if (friends !== "0") {
          var ccnt = 0
          for (let i=1; i<=3; i++) {
            ccnt += 1
            if (b[r+dr[i]][c+dc[i]] !== friends) {
              break
            }
            if (ccnt === 3) {
              cnt += 1
              for (let j=0; j<4; j++) {
                chk[r+dr[j]][c+dc[j]] = 1
              }
            }
          }
        }
      }
    }
  
    if (cnt === 0) {
      break
    }

    for (let r=0; r<m; r++) {
      for (let c=0; c<n; c++) {
        if (chk[r][c] === 1) {
          answer += 1
          for (let k=r-1; k>=0; k--) {
            b[k+1][c] = b[k][c]
          }
          b[0][c] = "0"
        }
      }
    }
  }

  return answer;
}

m = 5
n = 6
b = ["AAAAAA", "BBAATB", "BBAATB", "JJJTAA", "JJJTAA"]
console.log(solution(m, n, b))